open-craft/opencraft
# Update requirements and dependency badge

cf. [OC-2024](https://tasks.opencraft.com/browse/OC-2024)

**Notes**

* I could not update PyYAML, selenium, and Werkzeug to the most recent versions available, since Jasmine currently still depends on slightly older versions.

* https://github.com/tomchristie/django-rest-framework/pull/3273 changed the way breadcrumbs are generated for browsable API pages: Before, DRT would call a generic [`get_view_name`](https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/views.py#L25) function to obtain view names to use for individual breadcrumbs. That function only uses information that it can obtain from the view *class*, and does not depend on the `action` attribute being set on the view. Now, the code that generates breadcrumbs instantiates a view based on the view class and calls the (overridden) `get_view_name` function on it. This broke a couple of tests for the instance and app server APIs (`OpenEdXInstanceAPITestCase.test_view_name` and `OpenEdXAppServerAPITestCase.test_view_name`), because our custom implementations of `get_view_name` (`InstanceViewSet.get_view_name` and `OpenEdXAppServerViewSet.get_view_name`) depended on the `action` attribute being set. *DRT will only set the `.action` attribute on a view in the context of a request*, so when the breadcrumbs code called `get_view_name` on a newly instantiated `InstanceViewSet` or `OpenEdXAppServerViewSet`, that call would fail with `AttributeError: 'ViewSet' object has no attribute 'action'.`.

  faa6214450a9ad45163ca1ef4c4c6a8f67666780 fixes this issue by using `getattr` to retrieve the value of the `action` field, allowing the breadcrumbs code to finish successfully. However, a small problem remains: When generating breadcrumbs using the new method, DRT itself can not distinguish between list and detail views of an `InstanceViewSet`/`OpenEdXAppServerViewSet` (since `suffix` is `None`). This results in breadcrumbs being displayed as

  > `Api Root` / `Instance None` / `Instance None`

  on the details page of an instance. On `master` (which uses an older version of DRT that precedes https://github.com/tomchristie/django-rest-framework/pull/3273) we instead got:

  > `Api Root` / `Instance List` / `Instance Instance`

  It's possible that I'm missing something, but given that we can only use the value of the `action` attribute to determine an appropriate view name in the context of a request, I'm not sure how we could go about determining an appropriate view name in the context of breadcrumbs generation.

  There's always the option of sticking with an older version of DRT, or using a custom fork that reverts the changes from https://github.com/tomchristie/django-rest-framework/pull/3273; but at this point, that would mean using a significantly older version of DRT (3.3.3 vs. 3.5.3), or introducing additional overhead for maintaining a separate fork.

  **Update**: Custom `get_view_name` methods have been removed completely, cf. relevant discussion starting with https://github.com/open-craft/opencraft/pull/153#discussion_r87375404. Keeping this note for future reference.

**Test instructions**

* Observe green build on this PR.
* Run `make shell` locally to see that it starts up successfully.
* Run the development server. Observe that there are no deprecation warnings in the output. Check that you can access the instance manager in the browser (shouldn't raise a `DisallowedHost` exception). **Note** that if you run management commands with `PYTHONWARNINGS=all`, you'll still see some warnings, but they concern code installed into the venv; they are not triggered by instance manager code.

**Reviewers**
